cache/lru/lru_test.go

```diff
@@ -0,0 +1,224 @@
+	c.Put(4, &sizeable{value: 4, size: 6})
+	assertEqual(t, c.Len(), 1, "")
+	assertEqual(t, len(c.cache), 1, "")
+	four = getSizeableValue(c.Get(4))
+	assertEqual(t, four, 4, "")
+}
+
+// TestCacheFailsInsertionSizeBiggerCapacity tests that the cache fails the
+// put operation when the element's size is bigger than it's capacity.
+func TestCacheFailsInsertionSizeBiggerCapacity(t *testing.T) {
+	t.Parallel()
+	c := NewCache(2)
+
+	err := c.Put(1, &sizeable{value: 1, size: 3})
+	if err == nil {
+		t.Fatal("shouldn't be able to insert elements larger than cache")
+	}
+	assertEqual(t, c.Len(), 0, "")
+}
+
+// TestManySmallElementCanInsertAfterBigEviction tests that when a big element
+// is evicted from the Cache, multiple smaller ones can be inserted without an
+// eviction taking place.
+func TestManySmallElementCanInsertAfterBigEviction(t *testing.T) {
+	t.Parallel()
+	c := NewCache(3)
+
+	err := c.Put(1, &sizeable{value: 1, size: 3})
+	if err != nil {
+		t.Fatal("couldn't insert element")
+	}
+
+	assertEqual(t, c.Len(), 1, "")
+
+	c.Put(2, &sizeable{value: 2, size: 1})
+	two := getSizeableValue(c.Get(2))
+	oneEntry, _ := c.Get(1)
+	assertEqual(t, c.Len(), 1, "")
+	assertEqual(t, two, 2, "")
+	assertEqual(t, oneEntry, nil, "")
+
+	c.Put(3, &sizeable{value: 3, size: 1})
+	assertEqual(t, c.Len(), 2, "")
+
+	c.Put(4, &sizeable{value: 4, size: 1})
+	assertEqual(t, c.Len(), 3, "")
+
+	two = getSizeableValue(c.Get(2))
+	three := getSizeableValue(c.Get(3))
+	four := getSizeableValue(c.Get(4))
+	assertEqual(t, two, 2, "")
+	assertEqual(t, three, 3, "")
+	assertEqual(t, four, 4, "")
+}
+
+// TestReplacingElementValueSmallerSize tests that if an existing element is
+// replaced with a value of size smaller, that the size shrinks and we can
+// insert without an eviction taking place.
+func TestReplacingElementValueSmallerSize(t *testing.T) {
+	t.Parallel()
+	c := NewCache(2)
+
+	c.Put(1, &sizeable{value: 1, size: 2})
+
+	c.Put(1, &sizeable{value: 1, size: 1})
+	c.Put(2, &sizeable{value: 2, size: 1})
+	one := getSizeableValue(c.Get(1))
+	two := getSizeableValue(c.Get(2))
+	assertEqual(t, one, 1, "")
+	assertEqual(t, two, 2, "")
+	assertEqual(t, c.Len(), 2, "")
+}
+
+// TestReplacingElementValueBiggerSize tests that if an existing element is
+// replaced with a value of size bigger, that it evicts accordingly.
+func TestReplacingElementValueBiggerSize(t *testing.T) {
+	t.Parallel()
+	c := NewCache(2)
+
+	c.Put(1, &sizeable{value: 1, size: 1})
+	c.Put(2, &sizeable{value: 2, size: 1})
+
+	c.Put(1, &sizeable{value: 3, size: 2})
+	assertEqual(t, c.Len(), 1, "")
+	one := getSizeableValue(c.Get(1))
+	assertEqual(t, one, 3, "")
+}
+
+// TestConcurrency is a very simple test that checks concurrent access to the
+// lru cache. When running the test, "-race" option should be passed to
+// "go test" command.
+func TestConcurrency(t *testing.T) {
+	t.Parallel()
+	c := NewCache(5)
+
+	for i := 0; i < 5; i++ {
+		go func(i int) {
```

Comment:
Add all goroutines to a waitgroup, then wait for them all to be done before finishing the test.